Amber protein ff converter

proteneer opened the pull request:
This PR adds support for converting OpenMM's `amber99sbildn.xml` file into unique types. 

The output is a dictionary that looks like:

```
 ['ALA [(4, 5), (4, 6), (4, 7)]', 0.5],
```
Which denotes, for the ALA residue name, the **ordered** bonds [(4,5), (4,6), (4,7)] should be assigned using the same BCC with a value of 0.5 such that charge is transferred in the direction of 5->4, 6->4, and 7->4. (i.e. first element is incremented, second element is decremented)

Full result:
```[ ['ACE [(1, 0), (1, 2), (1, 3)]', 0.0],
                                    ['ACE [(1, 4)]', 0.0],
                                    ['ACE [(5, 4)]', 0.0],
                                    ['ALA [(0, 1)]', 0.0],
                                    ['ALA [(0, 2)]', 0.0],
                                    ['ALA [(2, 3)]', 0.0],
                                    ['ALA [(4, 2)]', 0.0],
                                    ['ALA [(2, 8)]', 0.0],
                                    ['ALA [(4, 5), (4, 6), (4, 7)]', 0.0],
                                    ['ALA [(9, 8)]', 0.0],
                                    ['ARG [(0, 1)]', 0.0],
                                    ['ARG [(2, 0)]', 0.0],
                                    ['ARG [(2, 3)]', 0.0],
                                    ['ARG [(4, 2)]', 0.0],
                                    ['ARG [(2, 22)]', 0.0],
                                    ['ARG [(4, 5), (4, 6)]', 0.0],
                                    ['ARG [(4, 7)]', 0.0],
                                    ['ARG [(8, 7), (9, 7)]', 0.0],
                                    ['ARG [(7, 10)]', 0.0],
                                    ['ARG [(10, 11), (10, 12)]', 0.0],
                                    ['ARG [(13, 10)]', 0.0],
                                    ['ARG [(13, 14)]', 0.0],
                                    ['ARG [(13, 15)]', 0.0],
                                    ['ARG [(16, 15), (19, 15)]', 0.0],
                                    ['ARG [(16, 17), (16, 18), (19, 20), (19, 21)]', 0.0],
                                    ['ARG [(23, 22)]', 0.0],
                                    ['ASN [(0, 1)]', 0.0],
                                    ['ASN [(0, 2)]', 0.0],
                                    ['ASN [(2, 3)]', 0.0],
                                    ['ASN [(4, 2)]', 0.0],
                                    ['ASN [(2, 12)]', 0.0],
                                    ['ASN [(4, 5), (4, 6)]', 0.0],
                                    ['ASN [(4, 7)]', 0.0],
                                    ['ASN [(8, 7)]', 0.0],
                                    ['ASN [(9, 7)]', 0.0],
                                    ['ASN [(9, 10), (9, 11)]', 0.0],
                                    ['ASN [(13, 12)]', 0.0],
                                    ['ASP [(0, 1)]', 0.0],
                                    ['ASP [(0, 2)]', 0.0],
                                    ['ASP [(2, 3)]', 0.0],
                                    ['ASP [(4, 2)]', 0.0],
                                    ['ASP [(2, 10)]', 0.0],
                                    ['ASP [(5, 4), (6, 4)]', 0.0],
                                    ['ASP [(4, 7)]', 0.0],
                                    ['ASP [(8, 7), (9, 7)]', 0.0],
                                    ['ASP [(11, 10)]', 0.0],
                                    ['CYM [(0, 1)]', 0.0],
                                    ['CYM [(2, 0)]', 0.0],
                                    ['CYM [(2, 3)]', 0.0],
                                    ['CYM [(2, 4)]', 0.0],
                                    ['CYM [(2, 8)]', 0.0],
                                    ['CYM [(4, 5), (4, 6)]', 0.0],
                                    ['CYM [(4, 7)]', 0.0],
                                    ['CYM [(9, 8)]', 0.0],
                                    ['CYS [(0, 1)]', 0.0],
                                    ['CYS [(0, 2)]', 0.0],
                                    ['CYS [(2, 3)]', 0.0],
                                    ['CYS [(4, 2)]', 0.0],
                                    ['CYS [(2, 9)]', 0.0],
                                    ['CYS [(4, 5), (4, 6)]', 0.0],
                                    ['CYS [(4, 7)]', 0.0],
                                    ['CYS [(7, 8)]', 0.0],
                                    ['CYS [(10, 9)]', 0.0],
                                    ['CYX [(0, 1)]', 0.0],
                                    ['CYX [(0, 2)]', 0.0],
                                    ['CYX [(2, 3)]', 0.0],
                                    ['CYX [(4, 2)]', 0.0],
                                    ['CYX [(2, 8)]', 0.0],
                                    ['CYX [(4, 5), (4, 6)]', 0.0],
                                    ['CYX [(4, 7)]', 0.0],
                                    ['CYX [(9, 8)]', 0.0],
                                    ['GLH [(0, 1)]', 0.0],
                                    ['GLH [(0, 2)]', 0.0],
                                    ['GLH [(2, 3)]', 0.0],
                                    ['GLH [(4, 2)]', 0.0],
                                    ['GLH [(2, 14)]', 0.0],
                                    ['GLH [(4, 5), (4, 6)]', 0.0],
                                    ['GLH [(4, 7)]', 0.0],
                                    ['GLH [(7, 8), (7, 9)]', 0.0],
                                    ['GLH [(7, 10)]', 0.0],
                                    ['GLH [(11, 10)]', 0.0],
                                    ['GLH [(12, 10)]', 0.0],
                                    ['GLH [(12, 13)]', 0.0],
                                    ['GLH [(15, 14)]', 0.0],
                                    ['GLN [(0, 1)]', 0.0],
                                    ['GLN [(2, 0)]', 0.0],
                                    ['GLN [(2, 3)]', 0.0],
                                    ['GLN [(2, 4)]', 0.0],
                                    ['GLN [(2, 15)]', 0.0],
                                    ['GLN [(4, 5), (4, 6)]', 0.0],
                                    ['GLN [(4, 7)]', 0.0],
                                    ['GLN [(7, 8), (7, 9)]', 0.0],
                                    ['GLN [(7, 10)]', 0.0],
                                    ['GLN [(11, 10)]', 0.0],
                                    ['GLN [(12, 10)]', 0.0],
                                    ['GLN [(12, 13), (12, 14)]', 0.0],
                                    ['GLN [(16, 15)]', 0.0],
                                    ['GLU [(0, 1)]', 0.0],
                                    ['GLU [(0, 2)]', 0.0],
                                    ['GLU [(2, 3)]', 0.0],
                                    ['GLU [(2, 4)]', 0.0],
                                    ['GLU [(2, 13)]', 0.0],
                                    ['GLU [(5, 4), (6, 4)]', 0.0],
                                    ['GLU [(7, 4)]', 0.0],
                                    ['GLU [(8, 7), (9, 7)]', 0.0],
                                    ['GLU [(7, 10)]', 0.0],
                                    ['GLU [(11, 10), (12, 10)]', 0.0],
                                    ['GLU [(14, 13)]', 0.0],
                                    ['GLY [(0, 1)]', 0.0],
                                    ['GLY [(2, 0)]', 0.0],
                                    ['GLY [(2, 3), (2, 4)]', 0.0],
                                    ['GLY [(2, 5)]', 0.0],
                                    ['GLY [(6, 5)]', 0.0],
                                    ['HID [(0, 1)]', 0.0],
                                    ['HID [(0, 2)]', 0.0],
                                    ['HID [(2, 3)]', 0.0],
                                    ['HID [(4, 2)]', 0.0],
                                    ['HID [(2, 15)]', 0.0],
                                    ['HID [(4, 5), (4, 6)]', 0.0],
                                    ['HID [(7, 4)]', 0.0],
                                    ['HID [(7, 8)]', 0.0],
                                    ['HID [(7, 13)]', 0.0],
                                    ['HID [(8, 9)]', 0.0],
                                    ['HID [(8, 10)]', 0.0],
                                    ['HID [(11, 10)]', 0.0],
                                    ['HID [(12, 10)]', 0.0],
                                    ['HID [(12, 13)]', 0.0],
                                    ['HID [(14, 13)]', 0.0],
                                    ['HID [(16, 15)]', 0.0],
                                    ['HIE [(0, 1)]', 0.0],
                                    ['HIE [(2, 0)]', 0.0],
                                    ['HIE [(2, 3)]', 0.0],
                                    ['HIE [(4, 2)]', 0.0],
                                    ['HIE [(2, 15)]', 0.0],
                                    ['HIE [(4, 5), (4, 6)]', 0.0],
                                    ['HIE [(4, 7)]', 0.0],
                                    ['HIE [(8, 7)]', 0.0],
                                    ['HIE [(13, 7)]', 0.0],
                                    ['HIE [(8, 9)]', 0.0],
                                    ['HIE [(10, 9)]', 0.0],
                                    ['HIE [(11, 9)]', 0.0],
                                    ['HIE [(11, 12)]', 0.0],
                                    ['HIE [(13, 11)]', 0.0],
                                    ['HIE [(13, 14)]', 0.0],
                                    ['HIE [(16, 15)]', 0.0],
                                    ['HIP [(0, 1)]', 0.0],
                                    ['HIP [(2, 0)]', 0.0],
                                    ['HIP [(2, 3)]', 0.0],
                                    ['HIP [(4, 2)]', 0.0],
                                    ['HIP [(2, 16)]', 0.0],
                                    ['HIP [(4, 5), (4, 6)]', 0.0],
                                    ['HIP [(7, 4)]', 0.0],
                                    ['HIP [(7, 8)]', 0.0],
                                    ['HIP [(7, 14)]', 0.0],
                                    ['HIP [(8, 9)]', 0.0],
                                    ['HIP [(10, 8)]', 0.0],
                                    ['HIP [(10, 11)]', 0.0],
                                    ['HIP [(10, 12)]', 0.0],
                                    ['HIP [(12, 13)]', 0.0],
                                    ['HIP [(14, 12)]', 0.0],
                                    ['HIP [(14, 15)]', 0.0],
                                    ['HIP [(17, 16)]', 0.0],
                                    ['ILE [(0, 1)]', 0.0],
                                    ['ILE [(2, 0)]', 0.0],
                                    ['ILE [(2, 3)]', 0.0],
                                    ['ILE [(2, 4)]', 0.0],
                                    ['ILE [(2, 17)]', 0.0],
                                    ['ILE [(5, 4)]', 0.0],
                                    ['ILE [(6, 4)]', 0.0],
                                    ['ILE [(10, 4)]', 0.0],
                                    ['ILE [(6, 7), (6, 8), (6, 9)]', 0.0],
                                    ['ILE [(10, 11), (10, 12)]', 0.0],
                                    ['ILE [(10, 13)]', 0.0],
                                    ['ILE [(13, 14), (13, 15), (13, 16)]', 0.0],
                                    ['ILE [(18, 17)]', 0.0],
                                    ['LEU [(0, 1)]', 0.0],
                                    ['LEU [(2, 0)]', 0.0],
                                    ['LEU [(2, 3)]', 0.0],
                                    ['LEU [(2, 4)]', 0.0],
                                    ['LEU [(2, 17)]', 0.0],
                                    ['LEU [(4, 5), (4, 6)]', 0.0],
                                    ['LEU [(4, 7)]', 0.0],
                                    ['LEU [(8, 7)]', 0.0],
                                    ['LEU [(9, 7), (13, 7)]', 0.0],
                                    ['LEU [(9, 10), (9, 11), (9, 12), (13, 14), (13, 15), (13, 16)]', 0.0],
                                    ['LEU [(18, 17)]', 0.0],
                                    ['LYS [(0, 1)]', 0.0],
                                    ['LYS [(2, 0)]', 0.0],
                                    ['LYS [(2, 3)]', 0.0],
                                    ['LYS [(4, 2)]', 0.0],
                                    ['LYS [(2, 20)]', 0.0],
                                    ['LYS [(4, 5), (4, 6)]', 0.0],
                                    ['LYS [(4, 7)]', 0.0],
                                    ['LYS [(8, 7), (9, 7)]', 0.0],
                                    ['LYS [(10, 7)]', 0.0],
                                    ['LYS [(10, 11), (10, 12)]', 0.0],
                                    ['LYS [(13, 10)]', 0.0],
                                    ['LYS [(13, 14), (13, 15)]', 0.0],
                                    ['LYS [(13, 16)]', 0.0],
                                    ['LYS [(16, 17), (16, 18), (16, 19)]', 0.0],
                                    ['LYS [(21, 20)]', 0.0],
                                    ['MET [(0, 1)]', 0.0],
                                    ['MET [(2, 0)]', 0.0],
                                    ['MET [(2, 3)]', 0.0],
                                    ['MET [(2, 4)]', 0.0],
                                    ['MET [(2, 15)]', 0.0],
                                    ['MET [(5, 4), (6, 4)]', 0.0],
                                    ['MET [(7, 4)]', 0.0],
                                    ['MET [(7, 8), (7, 9)]', 0.0],
                                    ['MET [(10, 7)]', 0.0],
                                    ['MET [(11, 10)]', 0.0],
                                    ['MET [(11, 12), (11, 13), (11, 14)]', 0.0],
                                    ['MET [(16, 15)]', 0.0],
                                    ['NME [(0, 1)]', 0.0],
                                    ['NME [(2, 0)]', 0.0],
                                    ['NME [(2, 3), (2, 4), (2, 5)]', 0.0],
                                    ['PHE [(0, 1)]', 0.0],
                                    ['PHE [(2, 0)]', 0.0],
                                    ['PHE [(2, 3)]', 0.0],
                                    ['PHE [(2, 4)]', 0.0],
                                    ['PHE [(2, 18)]', 0.0],
                                    ['PHE [(4, 5), (4, 6)]', 0.0],
                                    ['PHE [(4, 7)]', 0.0],
                                    ['PHE [(8, 7), (16, 7)]', 0.0],
                                    ['PHE [(8, 9), (16, 17)]', 0.0],
                                    ['PHE [(8, 10), (16, 14)]', 0.0],
                                    ['PHE [(10, 11), (14, 15)]', 0.0],
                                    ['PHE [(12, 10), (12, 14)]', 0.0],
                                    ['PHE [(12, 13)]', 0.0],
                                    ['PHE [(19, 18)]', 0.0],
                                    ['PRO [(0, 1)]', 0.0],
                                    ['PRO [(10, 0)]', 0.0],
                                    ['PRO [(1, 2), (1, 3)]', 0.0],
                                    ['PRO [(4, 1)]', 0.0],
                                    ['PRO [(4, 5), (4, 6)]', 0.0],
                                    ['PRO [(7, 4)]', 0.0],
                                    ['PRO [(7, 8), (7, 9)]', 0.0],
                                    ['PRO [(7, 10)]', 0.0],
                                    ['PRO [(10, 11)]', 0.0],
                                    ['PRO [(10, 12)]', 0.0],
                                    ['PRO [(13, 12)]', 0.0],
                                    ['SER [(0, 1)]', 0.0],
                                    ['SER [(2, 0)]', 0.0],
                                    ['SER [(2, 3)]', 0.0],
                                    ['SER [(2, 4)]', 0.0],
                                    ['SER [(2, 9)]', 0.0],
                                    ['SER [(5, 4), (6, 4)]', 0.0],
                                    ['SER [(7, 4)]', 0.0],
                                    ['SER [(7, 8)]', 0.0],
                                    ['SER [(10, 9)]', 0.0],
                                    ['THR [(0, 1)]', 0.0],
                                    ['THR [(2, 0)]', 0.0],
                                    ['THR [(2, 3)]', 0.0],
                                    ['THR [(2, 4)]', 0.0],
                                    ['THR [(2, 12)]', 0.0],
                                    ['THR [(5, 4)]', 0.0],
                                    ['THR [(6, 4)]', 0.0],
                                    ['THR [(10, 4)]', 0.0],
                                    ['THR [(6, 7), (6, 8), (6, 9)]', 0.0],
                                    ['THR [(10, 11)]', 0.0],
                                    ['THR [(13, 12)]', 0.0],
                                    ['TRP [(0, 1)]', 0.0],
                                    ['TRP [(2, 0)]', 0.0],
                                    ['TRP [(2, 3)]', 0.0],
                                    ['TRP [(4, 2)]', 0.0],
                                    ['TRP [(2, 22)]', 0.0],
                                    ['TRP [(4, 5), (4, 6)]', 0.0],
                                    ['TRP [(4, 7)]', 0.0],
                                    ['TRP [(7, 8)]', 0.0],
                                    ['TRP [(7, 21)]', 0.0],
                                    ['TRP [(8, 9)]', 0.0],
                                    ['TRP [(8, 10)]', 0.0],
                                    ['TRP [(10, 11)]', 0.0],
                                    ['TRP [(10, 12)]', 0.0],
                                    ['TRP [(13, 12)]', 0.0],
                                    ['TRP [(21, 12)]', 0.0],
                                    ['TRP [(13, 14)]', 0.0],
                                    ['TRP [(15, 13)]', 0.0],
                                    ['TRP [(15, 16)]', 0.0],
                                    ['TRP [(15, 17)]', 0.0],
                                    ['TRP [(17, 18)]', 0.0],
                                    ['TRP [(17, 19)]', 0.0],
                                    ['TRP [(19, 20)]', 0.0],
                                    ['TRP [(19, 21)]', 0.0],
                                    ['TRP [(23, 22)]', 0.0],
                                    ['TYR [(0, 1)]', 0.0],
                                    ['TYR [(2, 0)]', 0.0],
                                    ['TYR [(2, 3)]', 0.0],
                                    ['TYR [(2, 4)]', 0.0],
                                    ['TYR [(2, 19)]', 0.0],
                                    ['TYR [(4, 5), (4, 6)]', 0.0],
                                    ['TYR [(7, 4)]', 0.0],
                                    ['TYR [(7, 8), (7, 17)]', 0.0],
                                    ['TYR [(8, 9), (17, 18)]', 0.0],
                                    ['TYR [(8, 10), (17, 15)]', 0.0],
                                    ['TYR [(10, 11), (15, 16)]', 0.0],
                                    ['TYR [(10, 12), (15, 12)]', 0.0],
                                    ['TYR [(13, 12)]', 0.0],
                                    ['TYR [(13, 14)]', 0.0],
                                    ['TYR [(20, 19)]', 0.0],
                                    ['VAL [(0, 1)]', 0.0],
                                    ['VAL [(2, 0)]', 0.0],
                                    ['VAL [(2, 3)]', 0.0],
                                    ['VAL [(2, 4)]', 0.0],
                                    ['VAL [(2, 14)]', 0.0],
                                    ['VAL [(5, 4)]', 0.0],
                                    ['VAL [(6, 4), (10, 4)]', 0.0],
                                    ['VAL [(6, 7), (6, 8), (6, 9), (10, 11), (10, 12), (10, 13)]', 0.0],
                                    ['VAL [(15, 14)]', 0.0],
                                    ['HOH [(0, 1), (0, 2)]', 0.0]]}}
```

The example categorical classes look like:

![res_out_nonbonded](https://github.com/proteneer/timemachine/assets/2280724/d1995378-5670-4a42-b5f0-3d028e26b6aa)

maxentile commented on timemachine/ff/handlers/nonbonded.py:
```diff
@@ -506,6 +506,96 @@ class AM1BCCSolventHandler(AM1BCCHandler):
+            data, topology, residueTemplates, ignoreExternalBonds
+        )
+
+        total = 0
+        bond_idxs = []
+
+        # generate bcc deltas
+        bond_deltas = []
+        for index, val in enumerate(templateForResidue):
+            # get bonds within each residue
+            for src, dst in val.bonds:
+                bond_idxs.append([src + total, dst + total])
+
+                if (src, dst) in self.res_to_bonds_to_param_idxs[val.name]:
+                    delta = params[self.res_to_bonds_to_param_idxs[val.name][(src, dst)]]
+                    bond_deltas.append(delta)
+                elif (dst, src) in self.res_to_bonds_to_param_idxs[val.name]:
+                    # flip sign
+                    delta = params[self.res_to_bonds_to_param_idxs[val.name][(dst, src)]]
+                    bond_deltas.append(-delta)
+                else:
+                    assert 0
```
nit[performance / modularity]: might want to adjust this bit to compute an index array `param_idxs` instead of directly forming a list of `bond_deltas`?

so that we can split this into two functions: one that's a possibly expensive function of the topology alone (returning `initial_charges, bond_idxs, param_idxs`, contains lots of non-differentiable stuff)

and a simpler/cheaper/differentiable function of the parameters
`bond_deltas = params[param_idxs]; return apply_bond_charge_corrections(initial_charges, bond_idxs, bond_deltas, ...)`

(since we will call this multiple times with the same `topology` and varying `params`)

proteneer replied:
Resolved

proteneer commented:
Additions:

1. Critical bug fix: Forgot to do an extra level of indirection when going from a bond defined by `(src_atom, dst_atom)` to `(template_src_atom, template_dst_atom)`
2. Critical bug fix: Fixed typing for repeated residues (eg. ASP-ASP)
3. Critical bug fix: Treat HIS protonation states more robustly (enabling HIE/HID/HIP variations).
4. Added 2 tests, one to verify that symmetries are correct, another to verify that we can process the hif2a pdb file.

proteneer commented:
> nit[performance / modularity]: might want to adjust this bit to compute an index array param_idxs instead of directly forming a list of bond_deltas?

Done in https://github.com/proteneer/timemachine/pull/1197/commits/9484cfac90414a20acca44853241ba34398e1eab

jkausrelay commented on timemachine/ff/amber_converter.py:
```diff
@@ -0,0 +1,313 @@
+                        # this roughly generates the correct symmetries (probably at the cost of having more parameters)
+                        src_idx = int(cn.attributes["from"].value)
+                        dst_idx = int(cn.attributes["to"].value)
+                        src_type = atom_idxs_to_types[src_idx]
```
nit: The xml is always ordered Atoms -> Bonds ? Since this reads from the dict written to in the if Atom branch.

proteneer replied:
Yes - I'll add a comment for this.

jkausrelay commented on timemachine/ff/amber_converter.py:
```diff
@@ -0,0 +1,313 @@
+            bc_dict[bc].append(bil)
+
+        for v in bc_dict.values():
+            pattern = res_name + " " + str(v)
```
nit: Why not make this [res_name, v] ? i.e. a proper nested list?

proteneer replied:
Mostly to generate a "similar looking" list of parameters to our current format. I don't think our current format allows for nested lists, and it current expects a "flat list" of parameters.

jkausrelay commented on tests/test_handlers.py:
```diff
@@ -1037,3 +1042,154 @@ def test_harmonic_bonds_complete():
+    mol = make_residue_mol(atoms, bonds)
+    from rdkit.Chem import Draw
+
+    with open("debug.svg", "w") as fh:
+        svg = Draw.MolsToGridImage([mol], molsPerRow=4, useSVG=True)
+        fh.write(svg)
```
comment out

proteneer replied:
willfix

jkausrelay commented on tests/test_handlers.py:
```diff
@@ -1037,3 +1042,154 @@ def test_harmonic_bonds_complete():
+    for group in expected_symmetries:
+        first = group[0]
+        for other in group[1:]:
+            aae(bcc_charges[other], bcc_charges[first])
```
nit: idk what aae is

proteneer replied:
assert almost equal - will fix this

jkausrelay reviewed (approved):
Minor comments, LGTM
